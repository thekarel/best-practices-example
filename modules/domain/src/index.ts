export * from './cupcake/Cupcake'
export * from './cupcake/Feature'
export * from './cupcake/Kind'
export * from './cupcake/cupcakeFixture'
export * from './order/Order'
