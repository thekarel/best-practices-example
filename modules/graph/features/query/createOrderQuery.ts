import {gql} from 'apollo-server'

export const createOrderQuery = () => gql`
  mutation createOrder {
    createOrder(
      deliveryAddress: "New York"
      customerName: "Mr. Muffin"
      items: [
        {kind: sweet, features: [sugarFree], name: "One"}
        {kind: sweet, features: [vegan], name: "Two"}
        {kind: sweet, features: [exclusive], name: "Three"}
      ]
    ) {
      id
      customerName
      deliveryAddress
      items {
        name
      }
    }
  }
`
