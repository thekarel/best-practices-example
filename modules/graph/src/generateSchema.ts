import {AwilixContainer} from 'awilix'
import {buildSchemaSync} from 'type-graphql'
import {OrderResolver} from './order/OrderResolver'

export const generateSchema = (container: AwilixContainer) =>
  buildSchemaSync({
    resolvers: [OrderResolver],
    container: {
      get: (constructor) => container.build(constructor),
    },
  })
