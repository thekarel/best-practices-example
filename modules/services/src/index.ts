export * from './EmailNotification'
export * from './order/InvalidOrderPayloadError'
export * from './order/OrderService'
