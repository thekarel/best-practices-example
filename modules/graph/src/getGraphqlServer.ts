import {ApolloServer} from 'apollo-server'
import {AwilixContainer} from 'awilix'
import {generateSchema} from './generateSchema'

export const getGraphqlServer = (container: AwilixContainer): ApolloServer => {
  const schema = generateSchema(container)

  const server = new ApolloServer({
    schema,
  })

  return server
}
