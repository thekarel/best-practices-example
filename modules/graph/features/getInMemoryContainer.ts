import {EmailNotificationMock} from '@cupcake/notification'
import {OrderRepositoryMemory} from '@cupcake/repository'
import {OrderService} from '@cupcake/services'
import {asClass, createContainer} from 'awilix'
import {OrderResolver} from '../src/order/OrderResolver'

export const getInMemoryContainer = () =>
  createContainer({injectionMode: 'CLASSIC'}).register({
    emailNotification: asClass(EmailNotificationMock),
    orderRepository: asClass(OrderRepositoryMemory).singleton(),
    orderResolver: asClass(OrderResolver),
    orderService: asClass(OrderService),
  })
