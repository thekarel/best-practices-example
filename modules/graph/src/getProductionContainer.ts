import {EmailNotificationMock} from '@cupcake/notification'
import {OrderRepositoryMemory} from '@cupcake/repository'
import {OrderService} from '@cupcake/services'
import {asClass, createContainer} from 'awilix'
import {OrderResolver} from './order/OrderResolver'

export const getProductionContainer = () =>
  createContainer({injectionMode: 'CLASSIC'}).register({
    emailNotification: asClass(EmailNotificationMock),
    orderRepository: asClass(OrderRepositoryMemory).singleton(),
    orderResolver: asClass(OrderResolver),
    orderService: asClass(OrderService),
  })
