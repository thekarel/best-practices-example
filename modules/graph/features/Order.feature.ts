import 'reflect-metadata'
import {getGraphqlServer} from '../src/getGraphqlServer'
import {getInMemoryContainer} from './getInMemoryContainer'
import {createOrderQuery} from './query/createOrderQuery'
import {createTestClient, ApolloServerTestClient} from 'apollo-server-testing'

let testClient: ApolloServerTestClient

describe(`Order`, () => {
  beforeAll(async () => {
    const server = getGraphqlServer(getInMemoryContainer())
    testClient = createTestClient(server)
  })

  describe(`Create`, () => {
    it(`returns new order`, async () => {
      const response = await testClient.mutate({mutation: createOrderQuery()})

      expect(response).toMatchSnapshot({data: {createOrder: {id: expect.any(String)}}})
    })
  })
})
