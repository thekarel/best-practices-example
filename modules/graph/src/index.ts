import 'reflect-metadata'
import {getGraphqlServer} from './getGraphqlServer'
import {getProductionContainer} from './getProductionContainer'

const port = 8888
const server = getGraphqlServer(getProductionContainer())

console.log(`üßÅ GraphQL Playground running on http://localhost:${port}`)
server.listen(port)
