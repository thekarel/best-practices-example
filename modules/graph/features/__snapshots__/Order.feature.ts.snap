// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order Create returns new order 1`] = `
Object {
  "data": Object {
    "createOrder": Object {
      "customerName": "Mr. Muffin",
      "deliveryAddress": "New York",
      "id": Any<String>,
      "items": Array [
        Object {
          "name": "One",
        },
        Object {
          "name": "Two",
        },
        Object {
          "name": "Three",
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
